Feature: Add layer zoom state

srijitcoder opened the pull request:
### What is this PR about (Based on #248)

- This PR allows helps us to view zoom level state based on `min` and `max` zoom
- The color change state only visible when `showZoomLayerState` is set inside layer properties.
  <img src="https://github.com/EOX-A/EOxElements/assets/10809211/b404b66a-7791-49c1-a349-b40a4bb08510" width="75%" />
- The background color of layer change to `#f0f5f9` and `opacity` to `0.3` (From -> TO)
  <img src="https://github.com/EOX-A/EOxElements/assets/10809211/aba3a87f-39d0-46e8-a0a2-88d922d27f87" width="48%" /> <img src="https://github.com/EOX-A/EOxElements/assets/10809211/5a8f5820-d028-473c-bfe4-14467b58662a" width="48%" />



### Demo - [Loom link here](https://www.loom.com/share/ec82e250b58e4d3ba7a85c98788e8132)
  <a href="https://www.loom.com/share/ec82e250b58e4d3ba7a85c98788e8132"><img width="98%" src="https://cdn.loom.com/sessions/thumbnails/ec82e250b58e4d3ba7a85c98788e8132-with-play.gif"></a>

silvester-pari commented on elements/layercontrol/src/components/layer.js:
```diff
@@ -43,6 +52,21 @@ export class EOxLayerControlLayer extends LitElement {
+   */
+  updated() {
+    if (this.layer?.get("showZoomLayerState")) {
+      this.map.getView().on("change:resolution", () => {
```
Registering the listener for `change:resolution` each time the element updates (`updated()`), the listeners actually stack and the more the element it uses, the more time this re-rendering triggers.

Adding a console logs in e.g. lines 61 and 64 shows this issue, which seems to slow down the website after zooming back and forth a bit and changing changing layer order:

![image](https://github.com/EOX-A/EOxElements/assets/26576876/a0ca49e7-cf70-487d-a540-5519b7cb9270)
![image](https://github.com/EOX-A/EOxElements/assets/26576876/5278aa2a-968a-41fb-805f-6f0b86dafd56)

srijitcoder replied:
@silvester-pari 
I have resolved re-rendering and event stacking issue.
I have used `firstUpdated()` instead of `updated()` to stop stacking of event 

But it will still create event of layer which has `showZoomLayerState: true`. We can still use one event function by putting everything inside `layerList` and controlling it with a list of zoom layer states but then it won't make sense because the functionality should be part of `layer` not it's list.

silvester-pari reviewed (changes requested):
Thanks for this, it looks really promising, and we might be able to re-use a similar approach for other constraints as well, e.g. if a layer has a defined extent and the current map view is outside that extent. In that case the check could be something like `isLayerVisibleBasedOnZoomState && isLayerVisibleBasedOnExtent`. I have created https://github.com/EOX-A/EOxElements/issues/372 in order to track this.

If the stacking/loop of listeners as described in the comment is resolved, I think we can use this for the zoom for now and add the extent condition as described in the new issue at a later date.

silvester-pari reviewed (changes requested):
Thank you for the changes. This looks good, although I'd have three more suggestions for changing before merging:

- If possible, please undo the formatting changes in the markdown files. The CHANGELOG.md are created automatically from our release pipeline and I am not sure if this will case conflicts. We should somehow exclude the CHANGELOG.md from formatting alltogether, I added https://github.com/EOX-A/EOxElements/issues/401 in order to track this.
- Thanks for the example story, however I would suggest changing it a bit in order to make it more understandable. For example: only two layers (no group), WIND as background and regions as foreground, where regions has a `minZoom` of e.g. 5 and WIND has a `maxZoom` of e.g. 9. This would be an example of a real-world usecase, since vector layers (which slow down rendering) are usually hidden on lower zoom and raster layers (which get blurry or pixely) are usually hidden on higher zoom.
- instead of having `showZoomLayerState` on a layer level, I'd suggest having it as a configuration option for the entire layercontrol; my thought process is that if in some application the developer wants this functionality, it should be usually applied to all layers which have a minZoom and maxZoom. Having to declare this on every layer seems a bit verbose.

Also, it would be great if we could add a quick test case using [isVisible()](https://openlayers.org/en/latest/apidoc/module-ol_layer_Layer-Layer.html#isVisible).

srijitcoder commented:
> Thank you for the changes. This looks good, although I'd have three more suggestions for changing before merging:

I think all the points are straight forward. I will resolve these points first and then work on #371 changes.

srijitcoder commented:
@silvester-pari  All the points have been addressed in the second set of changes:

1. Reverted all changes in `CHANGELOG.md`.
2. Implemented example changes reflecting real scenarios - Included `minZoom: 2` for `Region` and `maxZoom: 9` for `WIND`.
3. Introduced `showZoomLayerState` to `eox-layercontrol`.
4. Incorporated test cases covering three scenarios:
  I. Default zoom level set to `1` (verifying `foo` is disabled).
  II. Zoom level adjusted to `3` (ensuring both layers are visible).
  III. Zoom level adjusted to `10` (verifying `bar` is disabled).

silvester-pari reviewed (approved):
Great contribution, thanks!
